engine/step/step.go

```diff
@@ -531,7 +531,11 @@ func PreRun(st *Step, values *values.Values, ss StateSetter, executedSteps map[s
 // AfterRun evaluates a step's "check" conditions after the Step's action has been performed
 // and impacts the entire task's execution flow through the provided StateSetter
 func AfterRun(st *Step, values *values.Values, ss StateSetter) {
-	if st.skipped || st.State == StateServerError || st.State == StateFatalError || st.ForEach != "" {
+	switch st.State {
```

Comment:
I believe an inverse condition that allows to continue the execution when the step is in a "final" state would be clearer, and probably would avoid an issue if a new intermediate/error state is introduced in the future,

Reply:
Right. I initially refrained from inverting this condition, specifying only "final" states, because custom states are considered "final" and we obviously can't list all custom states. Though, custom states are set by conditions, so it's a non-issue.
Still, if new "final" states are ever added, we would hit the same problem, wouldn't we?

Reply:
How about checking if the step state is not present in the slice `stepConditionValidStates`? This would resolve issues if new state is added in the future.

```suggestion
	if st.skipped || st.ForEach != "" || !slices.Contains(stepConditionValidStates, st.State) {
```

Reply:
How about checking if the step state is not present in slice `runnableStates`? This would resolve issues if new state is added in the future.

```suggestion
	if st.skipped || st.ForEach != "" || !slices.Contains(runnableStates, st.State) {
```

Reply:
We can't simply use `runnableStates` because some error states (like `CLIENT_ERROR`) are runnable but shouldn't prevent conditions from being executed. For instance, a step performing an HTTP call that returns a 4XX status code sets the step in the `CLIENT_ERROR` state, state that we would override in a condition because we expected to receive a 4XX.